### PR TITLE
ci: Update codecov action version

### PR DESCRIPTION
**Pull request by georgebisbas**
For checking whether this will help in dependabot issue

### CODE REVIEW

**Comment by georgebisbas** [comment]
Maybe related to this: https://github.com/codecov/codecov-action/blob/main/README.md


```md
Easily upload coverage reports to Codecov from GitHub Actions
v4 Release
v4 of the Codecov GitHub Action will use the [Codecov CLI](https://github.com/codecov/codecov-cli) to upload coverage reports to Codecov.

Breaking Changes
Tokenless uploading is unsupported. However, PRs made from forks to the upstream public repos will support tokenless (e.g. contributors to OS projects do not need the upstream repo's Codecov token)
Various arguments to the Action have been removed
Dependabot
For repositories using Dependabot, users will need to ensure that it has access to the Codecov token for PRs from Dependabot to upload coverage. To do this, please add your CODECOV_TOKEN as a Dependabot Secret. For more information, see ["Configuring access to private registries for Dependabot."](https://docs.github.com/en/code-security/dependabot/working-with-dependabot/configuring-access-to-private-registries-for-dependabot#storing-credentials-for-dependabot-to-use)
v3 versions and below will not have access to CLI features (e.g. global upload token, ATS).
```

**Comment by georgebisbas** [comment]
https://github.com/codecov/codecov-action/issues/1248

File: .github/workflows/ci-mlir.yml [hunk 1]
```diff
@@ -46,6 +46,7 @@ jobs:
         # Change directory so that xdsl-opt can be found during installation.
         cd xdsl
         pip install -r requirements.txt
+        pip install --upgrade coverage
```
Comment (superlopuh) [inline .github/workflows/ci-mlir.yml:49]: I'm surprised we need this, isn't coverage in requirements.txt? If not, let's add it, so everyone can run the CI locally

File: .github/workflows/ci-mlir.yml [hunk 2]
```diff
@@ -124,12 +124,12 @@ jobs:
...
+        action: codecov/codecov-action@v4
         attempt_delay: 10000
         attempt_limit: 10
         with: |
           fail_ci_if_error: true
           verbose: true
           directory: ${GITHUB_WORKSPACE}/../
           files: coverage.xml
-          token: ${{ secrets.CODECOV_TOKEN }}
+          token: ${{ secrets.CODECOV_TOKEN }}
```
Comment (superlopuh) [inline .github/workflows/ci-mlir.yml:135]:
```suggestion
          token: ${{ secrets.CODECOV_TOKEN }}

```

**Comment by superlopuh** [comment]
I have a couple of hunches, will push here to try

**Comment by superlopuh** [comment]
Yay :)

**Pull request merged**